features/370-auth-ui/design.md

```diff
@@ -0,0 +1,529 @@
+
+The signup page displays the first login ID key default. Other login ID keys are available to choose.
+
+```
+Sign up with email
+|--------------------------|  |----------|
+| Enter in your email here |  | Continue |
+|--------------------------|  |----------|
+
+Sign up with phone instead.
+Sign up with username instead.
+```
+
+### The signup password page
+
+The signup password page displays a visibility toggleable password field with password requirements.
+
+```
+|------------------------|  |----------|
+| Create a password here |  | Continue |
+|------------------------|  |----------|
+
+- [ ] At least one digit
+- [ ] At least one uppercase English character
+- [ ] At least one lowercase English character
+- [ ] At least one symbols ~`!@#$%^&*()-_=+[{]}\|;:'",<.>/?
+- [ ] At least 8 characters long
+```
+
+## The settings endpoint
+
+The settings endpoint `/settings` allows the user to manage their account. Features include but not limited to:
+
+- Update metadata
+- Change password
+- Configure MFA
+- Add / Update / remove login ID
+- Manage identities
+- Manage sessions
+
+## The verification endpoint
+
+The verification endpoint `/verify` allows the user to verify their login ID. This page requires authenticated user. Unauthenticated access is redirected to the authorization page. Verification is not enforced by the authentication flow. For example, if auto send verification email is enabled, the email is sent behind the scene and the user is redirected to the application immediately after authentication success.
+
+The verification email sent by Auth Gear includes this verification endpoint. The implication is that if the user has logged in Auth Gear but not Auth UI, they have to authenticate again to verify.
+
+## Omitted features
+
+Auth UI does not support authenticate with Custom Token.
+
+## Future improvements
+
+### Customizing the signup flow or the authentication flow
+
+Some customization the developer may want to make
+
+- Add inline extra fields in the signup page to fill in metadata
+- Add custom pages before the signup page to fill in metadata
+- Redirect to an external form page before the signup page and continue to the signup page with metadata
+- Add custom pages after authentication, and finally redirect the user back to the application.
+- Redirect to an external page after authentication, the external page redirects back to Auth UI, and finally redirect the user back to the application.
+
+Introducing the concept of [authentication flow](https://www.keycloak.org/docs/6.0/server_admin/#_authentication-flows) may fulfill some of the above use cases.
+
+## OpenID Provider conformance
+
+Some of the terms used in this session are defined in [OpenID Connect Core 1.0](https://openid.net/specs/openid-connect-core-1_0.html#Terminology).
+
+Auth UI implements the [Authorization Code Flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth) with [PKCE](https://tools.ietf.org/html/rfc7636)
```

Comment:
Totally random, but maybe we can bear in mind we may want to support more OpenID features in future, such as the standard claims (as metadata?), and other optional endpoints like UserInfo or Session management (actually for OpenID standard I think UserInfo is not optional btw)

Reply:
> Totally random, but maybe we can bear in mind we may want to support more OpenID features in future, such as the standard claims (as metadata?)

We DO want to support more OpenID features. In fact, [Terry considered standard claims](https://github.com/SkygearIO/features/blob/master/features/323-login-id/design.md#standard-keys). 

> actually for OpenID standard I think UserInfo is not optional btw

Correct. But Auth Gear does not need it at the moment. Implementing UserInfo also requires Authorization which we do not implement now.

Reply:
Nice, got another feature issue for it? I think we better cover it in another feature issue.

Reply:
Opened https://github.com/SkygearIO/features/issues/413